feat(BE): add specific table view

lshtar13 opened the pull request:
feat(BE): add specific table view

return all specific reservations according to client's request.
prisma에서 해당 예약 내역을 전달받으면, 이를 가공하여 예약 개수(total)와 상세 정보(data)로 구분하여 return 합니다.
(이 때, 약간의 가공이라는 것은 prisma에서 응답할 때 예약 정보가 저장되어있는 서식과 FE에서 원하는 서식이 불일치하기 때문에 이루어지는 작업으로 prisma는 예약 참여자들을 각각 개체로 만들어 반환하기 때문에 이를 .map()메서드를 활용하여 배열로 묶어주는 작업을 의미합니다.)

Close #40

lshtar13 committed:
Co-authored-by: Jisoo Park <jspark2000@users.noreply.github.com>

k1g99 commented on backend/src/reservation/reservation.controller.ts:
```diff
@@ -1,4 +1,15 @@
+    @Query('startTime') startTime: string,
+    @Query('endTime') endTime: string
+  ) {
+    console.log(startTime, endTime)
```
배포단계에서는 주석처리나 삭제가 필요할 것 같습니다!

lshtar13 replied:
옙, 수정하겠습니다!

k1g99 commented on backend/src/reservation/reservation.service.ts:
```diff
@@ -4,4 +4,56 @@ import { PrismaService } from 'src/prisma/prisma.service'
 export class ReservationService {
   constructor(private readonly prismaService: PrismaService) {}
+  async specificReservation(start: string, end: string) {
+    const dbresults = await this.prismaService.reservation.findMany({
```
파라미터로 전달되는 start와 end사이의 시간 간격이 30분으로 고정되어있는 것으로 이해했는데

예를 들어 `start = 2023-02-02T13:00:00` ~ `end = 2023-02-02T13:30:00` 으로 되어있고,
만약 저장된 일정이 `2023-02-02T13:00:00` ~ `2023-02-02T15:00:00` 으로 되어있다면, 
해당 일정을 쿼리하는 게 가능한가요??

(제가 잘못 이해했을 수도 있습니다;;;)

lshtar13 replied:
그 부분을 생각 못했네요;; 수정하겠습니다!

k1g99 reviewed (commented):
확인 부탁드립니다~!~!

lshtar13 committed:
console.log 삭제

console.log 삭제 하였습니다

lshtar13 committed:
Update Prisma Query

SH9480P commented:
PR 설명 작성해주세요~

SH9480P commented on backend/src/reservation/reservation.module.ts:
```diff
@@ -1,4 +1,4 @@
-import { Module } from '@nestjs/common'
+import { Get, Module } from '@nestjs/common'
```
Get 지워주세요

SH9480P commented on backend/src/reservation/reservation.controller.ts:
```diff
@@ -1,4 +1,14 @@
-export class ReservationController {}
+export class ReservationController {
+  constructor(private readonly reservationService: ReservationService) {}
+  @Get('detail')
+  getSpecificReservations(
+    @Query('startTime') startTime: string,
+    @Query('endTime') endTime: string
+  ) {
+    return this.reservationService.specificReservation(startTime, endTime)
+  }
+}
```
async, await 붙여주세요

SH9480P commented on backend/src/reservation/reservation.service.ts:
```diff
@@ -4,4 +4,58 @@ import { PrismaService } from 'src/prisma/prisma.service'
 @Injectable()
 export class ReservationService {
   constructor(private readonly prismaService: PrismaService) {}
+  async specificReservation(start: string, end: string) {
```
- `specificReservation` 메서드 반환값 타입 지정 해주세요!
- 함수 이름은 동사로 시작하는게 더 좋아보여요. getSpecificReservations은 어떨까요?

SH9480P commented on backend/src/reservation/reservation.service.ts:
```diff
@@ -4,4 +4,58 @@ import { PrismaService } from 'src/prisma/prisma.service'
 @Injectable()
 export class ReservationService {
   constructor(private readonly prismaService: PrismaService) {}
+  async specificReservation(start: string, end: string) {
+    const dbresults = await this.prismaService.reservation.findMany({
```
- 변수 이름은 camelCase로 해주세요.
- `dbresults` 말고 더 가독성 있는 이름은 어떨까요?

lshtar13 committed:
fix(be): add 'async-await'

add 'async-await'

lshtar13 committed:
fix(be): delete 'get'

delete 'get' method.

Close #43

lshtar13 committed:
fix(be): chage method and  value name

change 'dbresults' to 'rawData' and 'specificReservation' to 'getSpecificREservation'.

lshtar13 committed:
chore(be): rename some values and method and also delete unnecessary imported  mothod

rename 'specificReservation' to 'getSpecificReservation' and 'dbresults' to 'rawData' in reservation.service
delete 'Get' in reservation.module
attach 'async - await' to getSpecificReservation method in reservation.controller

lshtar13 committed:
Merge branch '40-feature-module-name-예약-상세-정보-가져오기' of https://github.com/skkuding/skkuzzim into 40-feature-module-name-예약-상세-정보-가져오기

SH9480P commented on backend/package.json:
```diff
@@ -26,6 +26,7 @@
     "@nestjs/core": "^9.0.0",
     "@nestjs/platform-express": "^9.0.0",
     "@prisma/client": "^4.8.1",
+    "pnpm": "^7.25.0",
     "reflect-metadata": "^0.1.13",
     "rxjs": "^7.2.0"
```
`npm install pnpm` 명령어로 설치하신 것 같네요.
`npm i -g pnpm`으로 설치하시고, package.json dependencies에 있는 pnpm은 지워주세요.

마찬가지로 package-lock.json도 npm install로 생성된 것 같네요. 
pnpm 사용 시 필요 없는 파일입니다. 지워주세요!

SH9480P commented on backend/src/reservation/reservation.controller.ts:
```diff
@@ -1,4 +1,14 @@
+  async getSpecificReservations(
+    @Query('startTime') startTime: string,
+    @Query('endTime') endTime: string
+  ) {
+    return await this.reservationService.specificReservation(startTime, endTime)
+  }
```
getSpecificReservation으로 변경해주세요~

SH9480P commented on backend/src/reservation/reservation.service.ts:
```diff
@@ -4,4 +4,58 @@ import { PrismaService } from 'src/prisma/prisma.service'
+          club: item.club,
+          startTime: item.startTime,
+          endTime: item.endTime,
+          purpose: item.purpose,
+          member: member
+        }
```
API 명세서대로 member's'로 바꿔주세요

SH9480P commented on backend/src/reservation/reservation.service.ts:
```diff
@@ -4,4 +4,58 @@ import { PrismaService } from 'src/prisma/prisma.service'
   constructor(private readonly prismaService: PrismaService) {}
+  async getSpecificReservation(start: string, end: string) {
+    const rawData = await this.prismaService.reservation.findMany({
+      where: {
+        OR: [
+          {
+            startTime: {
+              gte: new Date(start),
+              lte: new Date(end)
+            }
+          },
+          {
+            endTime: {
+              gte: new Date(start),
+              lte: new Date(end)
+            }
+          }
```
2시--2시 30분의 time block을 클릭했다고 가정해봅시다...
이 코드대로라면 예약의 startTime과 endTime이 2시와 2시 반 사이에 있는 예약 건만 조회되겠죠.
그러나 우리는 2시--2시 30분을 **포함하는 모든** 예약 건을 조회해야 합니다. 
이를테면, 1시에서 3시까지인 예약이 조회되어야 한다는 거죠.

SH9480P reviewed (changes requested):
이 부분만 변경해주시면 문제 없이 merge 가능할 것 같아요~ 화이팅~~

lshtar13 committed:
fix(be): fix prisma query

	 handle the case when both startTime and endTime exceed its request

	 Close #40

lshtar13 committed:
chore(be): prettier

	   settings

SH9480P commented on backend/src/reservation/reservation.service.ts:
```diff
@@ -4,4 +4,72 @@ import { PrismaService } from 'src/prisma/prisma.service'
 @Injectable()
 export class ReservationService {
   constructor(private readonly prismaService: PrismaService) {}
+  async getSpecificReservation(start: string, end: string) {
+    const rawData = await this.prismaService.reservation.findMany({
+      where: {
+        OR: [
+          {
+            startTime: {
+              gte: new Date(start),
+              lte: new Date(end)
+            }
+          },
+          {
+            endTime: {
+              gte: new Date(start),
+              lte: new Date(end)
+            }
+          },
+          {
+            AND: [
+              {
+                startTime: {
+                  lte: new Date(start)
+                }
+              },
+              {
+                endTime: {
+                  gte: new Date(end)
+                }
+              }
+            ]
+          }
+        ]
+      },
```
1시 ~ 1시 30분 조회하는 경우에 12 ~ 1시 예약이 조회되면 안되겠죠?
마찬가지로 1시 30분 ~ 3시 예약도 조회되면 안됩니다.
쿼리 수정해주세요~

SH9480P reviewed (changes requested):
# package-lock.json 삭제해주세요~~~

jspark2000 committed:
Merge remote-tracking branch 'origin/main' into 40-feature-module-name-예약-상세-정보-가져오기

jspark2000 committed:
feat(be): add class validation and dto

class validation과 DTO를 추가합니다.

Close #40

jspark2000 committed:
fix(be): delete package-lock.json

Close #40

jspark2000 committed:
fix(be): modify prisma query

선택한 시간을 포함하는 모든 예약 정보를 불러올 수 있도록 prisma 쿼리 코드를 수정했습니다.

Close #40

SH9480P reviewed (approved):
수고하셨습니다~ merge해주세요!
